snarker/src/external_snark_worker/external_snark_worker_effects.rs

```diff
@@ -0,0 +1,92 @@
+use std::sync::Arc;
+
+use mina_p2p_messages::v2::StateBodyHash;
+use shared::snark::Snark;
+
+use crate::{
+    external_snark_worker::ExternalSnarkWorkerPruneWorkAction,
+    snark_pool::{SnarkPoolAutoCreateCommitmentAction, SnarkPoolWorkAddAction},
+};
+
+use super::{ExternalSnarkWorkerAction, ExternalSnarkWorkerActionWithMeta, available_job_to_snark_worker_spec};
+
+pub fn external_snark_worker_effects<S: crate::Service>(
+    store: &mut crate::Store<S>,
+    action: ExternalSnarkWorkerActionWithMeta,
+) {
+    let (action, _) = action.split();
+    match action {
+        ExternalSnarkWorkerAction::Start(_) => {
+            let Some(path) = store.state().config.path.as_ref().cloned() else {
+                return;
+            };
+            if let Err(err) = store.service().start(path) {
+                todo!("report error {err:?}");
+            }
+        }
+        ExternalSnarkWorkerAction::Started(_) => {
+            store.dispatch(SnarkPoolAutoCreateCommitmentAction {});
+        }
+        ExternalSnarkWorkerAction::Kill(_) => {
+            if let Err(err) = store.service().kill() {
+                todo!("report error {err:?}");
+            }
+        }
+        ExternalSnarkWorkerAction::Killed(_) => {}
+        ExternalSnarkWorkerAction::Error(action) => {
+            todo!("report {err:?}", err = action.error);
+        }
+        ExternalSnarkWorkerAction::SubmitWork(action) => {
+            let job_id = &action.job_id;
+            let config = &store.state().config;
+            let public_key = config.public_key.clone();
+            let fee = config.fee.clone();
+            let Some(job) = store.state().snark_pool.get(job_id) else {
+                return;
+            };
+            let protocol_state_body = |block_hash: StateBodyHash| {
+                store
```

Comment:
We should try to find protocol state here as well: https://github.com/openmina/openmina/blob/7a23546e3fadb33de4d0c49c13668539c8442568/snarker/src/transition_frontier/transition_frontier_state.rs#L16

Job might refer to block older than our root. In such case protocol state will be there.

Reply:
I've replaced the closure with a fn in the transition frontier. BTW, we should have a map from state body hashes, to avoid recalculating them each time...